Mark crashing tests as known failures

rth opened the pull request:
This marks crashing tests as known failures in pytest,
 - the  crashing tests will still run, but will not error at a failure. The advantage is that if due to some unrelated reason (e.g. emsdk version update) it starts to pass, it could could be seen as XPASS in the pytest output. The pytest output also becomes more complete, and  `XFAIL` tests can be seen as a list that potentially might need some investigation.
 - added ability to mark as known failure python core tests that fail exclusively on chrome or firefox (in particular https://github.com/iodide-project/pyodide/issues/108) via `crash-chrome` and `crash-firefox` flags in `test/python_tests.txt`.
 - I'm not fully sure if the current separation between skipped and marked as failing tests in `python_tests.txt` makes sense.
 - marking pandas tests as a known failure on Chrome.
 - show a report of skipped and XFAIL tests at the end of the pytest session with `-r sxX`

The overall goal of this is to make CI green on master, so that in new PRs a CI failure would actually mean some issue with the PR.

rth commented on test/test_pandas.py:
```diff
@@ -1,13 +1,18 @@
-@pytest.mark.skip
-def test_extra_import(selenium):
+def test_extra_import(selenium, request):
+    if 'chrome' in selenium.__class__.__name__.lower():
```
This would be needed quite often in tests, I'm not sure if there is better solution than  monkeypatching `selenium.webdriver.{Firefox,Chrome}` with the `name` attribute or similar in `tests/conftest.py`, eg..
```py
class ChromeWrapper(SeleniumWrapper):
    def get_driver(self):
        [...]
        driver = Chrome(chrome_options=options)
        driver.name = 'chrome'
        return driver
```
?

mdboom replied:
Since they are in pyodide, I think `ChromeWrapper` and `FirefoxWrapper` could both just add a name attribute we can check against, e.g.:

```
class ChromeWrapper(SeleniumWrapper):
    name = 'chrome'
```

and then:

```
if selenium.name == 'chrome':
```

Maybe `browser` is a better name than `name`...

rth commented:
Thanks for the feedback @mdboom ! I addressed your comment. 

Also, I was wondering is whether you frequently saw Circle CI timeouts due to 10min without output. I restarted the last commit (before your review) 3 times (cf first 3 builds in https://circleci.com/gh/iodide-project/pyodide/tree/pull%2F117), and each time tests timed-out at different stages on the test suite around 40-50% of completion. I wonder if it could be due to this PR or if it's just bad luck (I haven't seen this in other PRs though).  Rebased on master, waiting to see if tests will finish this time.

Another nice point about this, is that it shows that some CPython tests marked as crashing actually pass (maybe since emsdk update), cf ouput in https://circleci.com/gh/iodide-project/pyodide/231 . A have therefore removed the `crash` flag from those in a25c516

mdboom commented:
I have seen those timeouts elsewhere occasionally.  I'm not sure what causes them, because it doesn't seem to be related to any particular test.  I'll file a new issue for this.

rth commented:
This time tests finished (and passed except for 2 due to a typo on my part). Also in total there are now 19 CPython tests previously marked as crashing that are passing, which is nice.

mdboom commented:
In the latest test run, it seems to get to about 80% of the way done and then all the chrome tests fail afterward.  I think this is related to merging the singleton selenium driver change into master which is now being picked up here.  I wonder if we can hook into the rerun functionality such that if a test is rerun, it will automatically get a new Selenium driver instance...

rth commented:
Hmm, yes, somehow the  known failed tests modify the environment in a way that make other tests fail..

> I wonder if we can hook into the rerun functionality such that if a test is rerun, it will automatically get a new Selenium driver instance...

I agreed that ideally one would want to be able to select which CPython test to run in a clean selenium instance. There are approaches like https://github.com/pytest-dev/pytest/issues/2424#issuecomment-376012022 but it seems overly complicated / brittle. I have not found a simple solution yet..

rth commented:
More pragmatically, now crashing CPython tests are marked as `XFAIL` without executing them. So it is one step back, but the overall mechanism is still included and https://github.com/iodide-project/pyodide/pull/117/commits/e2a723de5f2e0f071ab604b6126873be3225faa1 could be easily reverted in the future when we find which `XFAIL` tests change  the global environment and make other tests fail.

Also disabled `pytest-rerunfailures` as I'm hoping this won't be necessary anymore, it makes the log less nice when tests are genuinely failing (e.g. in a PR) and their [readme](https://github.com/pytest-dev/pytest-rerunfailures) mentions that it's not compatible with module level fixture (which we are indirectly using).

In any case CI is finally green, so I think this is ready for a final review.

mdboom commented:
It's green!!!  I like the idea of automatically knowing when failing tests are now passing, but it's also possible to just reexamine them locally from time to time when major changes, such us updating Python, occur.  

I'm going to merge this and make a new issue to re-examine the global effects issue later.

rth commented:
> In the latest test run, it seems to get to about 80% of the way done and then all the chrome tests fail afterward. 

The general situation is that after a segfault, all the following tests will fail. Re-running the test suite with all tests enabled and `ASSERTIONS=2`, and so far I'm getting,
 - syscall3 / read : segfaults in `test_io`
 - syscall269 / fstatfs64 : segfaults in `test_posix`

Looking for more..
